IMPORT LIBRARY

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
#%matplotlib inline
import seaborn as sns

IMPORT FILE CSV

In [2]:
df = pd.read_csv('ds_salaries.csv')

In [3]:
df.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

# 1 Exploratory Data Analysis




## 1.1 Check Null Values

In [4]:
df.isna().sum()
#Semua kolom terisi dengan sempurna,
#tidak ada data yang kosong dari semua kolom

work_year             0
experience_level      0
employment_type       0
job_title             0
salary                0
salary_currency       0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
dtype: int64

## 1.2 Check Duplicate Data

In [5]:
df.shape

(3755, 11)

In [6]:
df[df.duplicated()]
#Ada 1171 data yang duplicated

work_year experience_level employment_type           job_title  salary  \
115        2023               SE              FT      Data Scientist  150000   
123        2023               SE              FT  Analytics Engineer  289800   
153        2023               MI              FT       Data Engineer  100000   
154        2023               MI              FT       Data Engineer   70000   
160        2023               SE              FT       Data Engineer  115000   
...         ...              ...             ...                 ...     ...   
3439       2022               MI              FT      Data Scientist   78000   
3440       2022               SE              FT       Data Engineer  135000   
3441       2022               SE              FT       Data Engineer  115000   
3586       2021               MI              FT       Data Engineer  200000   
3709       2021               MI              FT      Data Scientist   76760   

     salary_currency  salary_in_usd employee_residence  remote_ratio  \
115              USD         150000                 US             0   
123              USD         289800                 US             0   
153              USD         100000                 US           100   
154              USD          70000                 US           100   
160              USD         115000                 US             0   
...              ...            ...                ...           ...   
3439             USD          78000                 US           100   
3440             USD         135000                 US           100   
3441             USD         115000                 US           100   
3586             USD         200000                 US           100   
3709             EUR          90734                 DE            50   

     company_location company_size  
115                US            M  
123                US            M  
153                US            M  
154                US            M  
160                US            M  
...               ...          ...  
3439               US            M  
3440               US            M  
3441               US            M  
3586               US            L  
3709               DE            L  

[1171 rows x 11 columns]

In [7]:
df[df.duplicated()].shape

(1171, 11)

### Copy Data

In [8]:
df_c = df.copy()

In [9]:
df_c.head()

work_year experience_level employment_type                 job_title  \
0       2023               SE              FT  Principal Data Scientist   
1       2023               MI              CT               ML Engineer   
2       2023               MI              CT               ML Engineer   
3       2023               SE              FT            Data Scientist   
4       2023               SE              FT            Data Scientist   

   salary salary_currency  salary_in_usd employee_residence  remote_ratio  \
0   80000             EUR          85847                 ES           100   
1   30000             USD          30000                 US           100   
2   25500             USD          25500                 US           100   
3  175000             USD         175000                 CA           100   
4  120000             USD         120000                 CA           100   

  company_location company_size  
0               ES            L  
1               US            S  
2               US            S  
3               CA            M  
4               CA            M

In [10]:
df_c.shape

(3755, 11)

### Delete Duplicate Data

In [11]:
df_c = df_c.drop_duplicates()

In [12]:
df_c.shape

(2584, 11)

## 1.3 Check Unique Values in Each Column

### Check Data Type

In [13]:
df_c.dtypes

work_year              int64
experience_level      object
employment_type       object
job_title             object
salary                 int64
salary_currency       object
salary_in_usd          int64
employee_residence    object
remote_ratio           int64
company_location      object
company_size          object
dtype: object

### Check Data stats

In [14]:
df_c.describe()

work_year        salary  salary_in_usd  remote_ratio
count  2584.000000  2.584000e+03    2584.000000   2584.000000
mean   2022.301084  2.103653e+05  133409.280186     50.483746
std       0.749179  8.080375e+05   67136.837329     48.163707
min    2020.000000  6.000000e+03    5132.000000      0.000000
25%    2022.000000  9.000000e+04   84975.000000      0.000000
50%    2022.000000  1.346300e+05  130000.000000     50.000000
75%    2023.000000  1.825625e+05  175000.000000    100.000000
max    2023.000000  3.040000e+07  450000.000000    100.000000

### Check Columns

In [15]:
df_c.columns

Index(['work_year', 'experience_level', 'employment_type', 'job_title',
       'salary', 'salary_currency', 'salary_in_usd', 'employee_residence',
       'remote_ratio', 'company_location', 'company_size'],
      dtype='object')

### Unique Valeus Check

In [16]:
for column in df_c.columns:
  if df_c[column].dtypes == object:
    num_distinct = df_c[column].nunique() # Ada berapa unique values
    distinct_val = df_c[column].unique() # Unique valuenya apa saja

    print(f"{column} : {num_distinct} --> {distinct_val}\n")


experience_level : 4 --> ['SE' 'MI' 'EN' 'EX']

employment_type : 4 --> ['FT' 'CT' 'FL' 'PT']

job_title : 93 --> ['Principal Data Scientist' 'ML Engineer' 'Data Scientist'
 'Applied Scientist' 'Data Analyst' 'Data Modeler' 'Research Engineer'
 'Analytics Engineer' 'Business Intelligence Engineer'
 'Machine Learning Engineer' 'Data Strategist' 'Data Engineer'
 'Computer Vision Engineer' 'Data Quality Analyst'
 'Compliance Data Analyst' 'Data Architect'
 'Applied Machine Learning Engineer' 'AI Developer' 'Research Scientist'
 'Data Analytics Manager' 'Business Data Analyst' 'Applied Data Scientist'
 'Staff Data Analyst' 'ETL Engineer' 'Data DevOps Engineer' 'Head of Data'
 'Data Science Manager' 'Data Manager' 'Machine Learning Researcher'
 'Big Data Engineer' 'Data Specialist' 'Lead Data Analyst'
 'BI Data Engineer' 'Director of Data Science'
 'Machine Learning Scientist' 'MLOps Engineer' 'AI Scientist'
 'Autonomous Vehicle Technician' 'Applied Machine Learning Scientist'
 'Lead Data S

## 1.4 Rename Value for Better Understanding

In [17]:
df_c['experience_level'] = df_c['experience_level'].replace({
    'SE' : 'Senior',
    'EN' : 'Entry Level',
    'EX' : 'Executive Level',
    'MI' : 'Mid/Intermidiate Level'
})

df_c['employment_type'] = df_c['employment_type'].replace({
    'FT' : 'Full-time',
    'PT' : 'Part-time',
    "FL" : 'Freelance',
    'CT' : 'Contractor'
})

df_c['remote_ration'] = df_c['remote_ratio'].replace({
    0 : 'On-Site',
    50: 'Hybrid',
    100: 'Full-Remote'
})

## 1.5 Group the Job title

### Check Job Title

In [18]:
print(df_c['job_title'].nunique())
print(df_c['job_title'].unique())

93
['Principal Data Scientist' 'ML Engineer' 'Data Scientist'
 'Applied Scientist' 'Data Analyst' 'Data Modeler' 'Research Engineer'
 'Analytics Engineer' 'Business Intelligence Engineer'
 'Machine Learning Engineer' 'Data Strategist' 'Data Engineer'
 'Computer Vision Engineer' 'Data Quality Analyst'
 'Compliance Data Analyst' 'Data Architect'
 'Applied Machine Learning Engineer' 'AI Developer' 'Research Scientist'
 'Data Analytics Manager' 'Business Data Analyst' 'Applied Data Scientist'
 'Staff Data Analyst' 'ETL Engineer' 'Data DevOps Engineer' 'Head of Data'
 'Data Science Manager' 'Data Manager' 'Machine Learning Researcher'
 'Big Data Engineer' 'Data Specialist' 'Lead Data Analyst'
 'BI Data Engineer' 'Director of Data Science'
 'Machine Learning Scientist' 'MLOps Engineer' 'AI Scientist'
 'Autonomous Vehicle Technician' 'Applied Machine Learning Scientist'
 'Lead Data Scientist' 'Cloud Database Engineer' 'Financial Data Analyst'
 'Data Infrastructure Engineer' 'Software Data Eng

### Grouping Function

In [19]:
def assign_broader_category(job_title):
  data_engineering = ["Data Engineer", "Data Analyst", "Analytics Engineer", "BI Data Analyst", "Business Data Analyst", "BI Developer", "BI Analyst", "Business Intelligence Engineer", "BI Data Engineer", "Power BI Developer"]
  data_scientist = ["Data Scientist", "Applied Scientist", "Research Scientist", "3D Computer Vision Researcher", "Deep Learning Researcher", "AI/Computer Vision Engineer"]
  machine_learning = ["Machine Learning Engineer", "ML Engineer", "Lead Machine Learning Engineer", "Principal Machine Learning Engineer"]
  data_architecture = ["Data Architect", "Big Data Architect", "Cloud Data Architect", "Principal Data Architect"]
  management = ["Data Science Manager", "Director of Data Science", "Head of Data Science", "Data Scientist Lead", "Head of Machine Learning", "Manager Data Management", "Data Analytics Manager"]

  if job_title in data_engineering:
    return "Data Engineering"
  elif job_title in data_scientist:
    return "Data Scientist"
  elif job_title in data_architecture:
    return "Data Architecture"
  elif job_title in machine_learning:
    return "Machine Learning"
  elif job_title in management:
    return "Management"
  else:
    return "Other"

In [20]:
df_c['job_category'] = df_c['job_title'].apply(assign_broader_category)

In [21]:
print(df_c['job_category'].nunique())
print(df_c['job_category'].unique())

6
['Other' 'Machine Learning' 'Data Scientist' 'Data Engineering'
 'Data Architecture' 'Management']


## 1.6 Adjust Salary to Present Values

In [22]:
df_c.columns

Index(['work_year', 'experience_level', 'employment_type', 'job_title',
       'salary', 'salary_currency', 'salary_in_usd', 'employee_residence',
       'remote_ratio', 'company_location', 'company_size', 'remote_ration',
       'job_category'],
      dtype='object')

### Inflation Rate

In [23]:
# Inflation rates
us_inflation_rates = {2019: 0.0181,
                      2020: 0.0123,
                      2021: 0.0470,
                      2022: 0.065}
global_inflation_rates = {2019: 0.0219,
                          2020: 0.0192,
                          2021: 0.0350,
                          2022: 0.088}

def adjust_salary(x):
  year = x['work_year']
  ori_salary = x['salary_in_usd']
  currency = x['salary_currency']

  # Kondisi perama (Jika th 2023 maka tidak terkena inflasi)
  if year == 2023:
    return ori_salary
  else:
    adjust_salary = ori_salary
    for y in range(year,2023):# Loop tahun sampai tahun 2022/ stop sebelum tahun 2023
      # Kondisi kedua (Jika currency adalah USD maka digunakan us_inflation rate, selain USD gunakan global )
      if currency == 'USD':
        inflation_rate = us_inflation_rates[y]
      else:
        inflation_rate = global_inflation_rates[y]

      adjust_salary = adjust_salary * (1+inflation_rate)
    return adjust_salary

In [24]:
df_c['adjusted_salary'] = df_c.apply(adjust_salary, axis=1)

In [25]:
df_c.tail()

work_year        experience_level employment_type  \
3750       2020                  Senior       Full-time   
3751       2021  Mid/Intermidiate Level       Full-time   
3752       2020             Entry Level       Full-time   
3753       2020             Entry Level      Contractor   
3754       2021                  Senior       Full-time   

                     job_title   salary salary_currency  salary_in_usd  \
3750            Data Scientist   412000             USD         412000   
3751  Principal Data Scientist   151000             USD         151000   
3752            Data Scientist   105000             USD         105000   
3753     Business Data Analyst   100000             USD         100000   
3754      Data Science Manager  7000000             INR          94665   

     employee_residence  remote_ratio company_location company_size  \
3750                 US           100               US            L   
3751                 US           100               US            L   
3752                 US           100               US            S   
3753                 US           100               US            L   
3754                 IN            50               IN            L   

     remote_ration      job_category  adjusted_salary  
3750   Full-Remote    Data Scientist    465053.312718  
3751   Full-Remote             Other    168373.305000  
3752   Full-Remote    Data Scientist    118520.868532  
3753   Full-Remote  Data Engineering    112877.017650  
3754        Hybrid        Management    106600.363200

# SOAL No.1
Show the median adjusted data scientist salary, for each experience level

## JAWABAN No.1

In [26]:
median_salary_by_experience = df.groupby('experience_level')['salary_in_usd'].median().reset_index()
median_salary_by_experience.columns = ['experience_level', 'median_salary_in_usd']

# Convert the result to a DataFrame
median_salary_by_experience_df = pd.DataFrame(median_salary_by_experience)

# Display the result
median_salary_by_experience_df

experience_level  median_salary_in_usd
0               EN               70000.0
1               EX              196000.0
2               MI              100000.0
3               SE              146000.0

Gaji data scientist secara signifikan meningkat dengan level pengalaman. Data ini menunjukkan pentingnya pengalaman dalam menentukan tingkat kompensasi di industri data science. Gaji tertinggi dicapai pada level Experienced (EX), yang mungkin mencerminkan peran spesialisasi yang sangat tinggi.

# SOAL No. 2
Show the number of records for each company_location, sorted descending

## JAWABAN No.2

In [27]:
records_by_location = df['company_location'].value_counts().reset_index()
records_by_location.columns = ['company_location', 'num_records']

# Sort the result in descending order
records_by_location = records_by_location.sort_values(by='num_records', ascending=False)

# Convert the result to a DataFrame
records_by_location_df = pd.DataFrame(records_by_location)

# Display the result
records_by_location_df

company_location  num_records
0                US         3040
1                GB          172
2                CA           87
3                ES           77
4                IN           58
..              ...          ...
52               NZ            1
51               CL            1
50               MD            1
49               EG            1
71               MT            1

[72 rows x 2 columns]

Mayoritas data berasal dari perusahaan yang berlokasi di Amerika Serikat, menunjukkan bahwa pasar data scientist di sana mungkin lebih besar atau lebih mudah diakses data gajinya. Jumlah rekor yang lebih sedikit dari negara lain bisa disebabkan oleh berbagai faktor seperti ukuran pasar kerja, kebijakan perusahaan, atau transparansi data gaji.

# SOAL No.3
From this data, show the median adjusted salary of employee who works for company located in Indonesia, do you think that value represents the salary in Indonesia? why? why not?

## JAWABAN No.3

In [28]:
indonesia_salaries = df[df['company_location'] == 'ID']['salary_in_usd']

# Calculate the median salary
median_indonesia_salary = indonesia_salaries.median()

print(f"Median adjusted salary in Indonesia: ${median_indonesia_salary}")

Median adjusted salary in Indonesia: $34208.0


**Kesimpulan** <br>
Gaji median sebesar $34,208 untuk data scientist di Indonesia atau sekita Rp 500.000.000, kemungkinan besar tidak sepenuhnya mencerminkan gaji lokal di Indonesia.  Hali ini dikarenakan ada beberapa faktor yang perlu dipertimbangkan yaitu:

1. Penyesuaian Gaji: "Adjusted salary" berarti gaji telah disesuaikan untuk faktor-faktor tertentu seperti biaya hidup atau nilai tukar mata uang, yang dapat meningkatkan atau menurunkan nilai nominal.

2. Perusahaan Internasional: Banyak perusahaan yang berlokasi di Indonesia mungkin adalah cabang dari perusahaan multinasional yang membayar gaji dengan standar internasional, lebih tinggi dari perusahaan lokal.

3. Variasi Gaji Lokal: Gaji lokal bisa sangat bervariasi tergantung industri, perusahaan, dan lokasi spesifik dalam negara.

Oleh karena itu, meskipun gaji median yang disesuaikan memberikan gambaran, mungkin tidak sepenuhnya akurat mewakili realitas gaji data scientist yang bekerja di perusahaan lokal Indonesia.